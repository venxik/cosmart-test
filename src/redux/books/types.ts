export type WorkDetails = {
  key: string;
  title: string;
  edition_count: number;
  authors: AuthorsDetails[];
  has_fulltext: boolean;
  ia: string;
};

export type AuthorsDetails = {
  name: string;
  key: string;
};

export type SubjectsResponse = {
  key: string;
  name: string;
  subject_type: string;
  work_count: number;
  works: WorkDetails[];
};

export type BookingInformation = WorkDetails & { bookingDate: string };

export type InitialState = {
  bookingInformation: BookingInformation[];
};
