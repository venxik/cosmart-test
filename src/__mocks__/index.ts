export * from './wrapper';
export * from './responseMock';
