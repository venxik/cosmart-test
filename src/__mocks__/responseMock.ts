export const getBooksResponseMock = {
  key: '/subjects/love',
  name: 'love',
  subject_type: 'subject',
  work_count: 15853,
  works: [
    {
      authors: [{ key: '/authors/OL4327048A', name: 'Emily BrontÃ«' }],
      key: 'testing',
      title: 'testing',
      edition_count: 123,
      has_fulltext: true,
      ia: 'testing',
      bookingDate: '1234',
    },
  ],
};
