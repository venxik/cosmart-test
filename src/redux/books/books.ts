import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { BookingInformation, InitialState } from './types';

const initialState: InitialState = {
  bookingInformation: [],
};

const reducerName = 'books';

const slice = createSlice({
  name: reducerName,
  initialState,
  reducers: {
    addBookingInformation(state: InitialState, action: PayloadAction<BookingInformation>) {
      state.bookingInformation = [...state.bookingInformation, action.payload];
    },
  },
});

export const booksReducer = { [reducerName]: slice.reducer };

export const { addBookingInformation } = slice.actions;
