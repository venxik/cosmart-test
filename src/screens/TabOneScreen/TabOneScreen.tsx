import { StyleSheet } from 'react-native';

import { View, BookList } from '../../components';

export default function TabOneScreen() {
  return (
    <View style={styles.container}>
      <BookList />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
