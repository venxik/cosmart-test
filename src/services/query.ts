import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';
import { SubjectsResponse } from '../redux/books/types';

const reducerPath = 'booksapi';
export const booksApi = createApi({
  reducerPath,
  keepUnusedDataFor: process.env.NODE_ENV === 'test' ? 0 : 60,
  baseQuery: fetchBaseQuery({ baseUrl: 'https://openlibrary.org/' }),
  endpoints: (builder) => ({
    getLoveSubjects: builder.query<SubjectsResponse, void>({
      query: () => ({
        url: 'subjects/love.json',
        method: 'GET',
      }),
    }),
  }),
});

export const { useGetLoveSubjectsQuery } = booksApi;

export const booksQueryReducer = { [reducerPath]: booksApi.reducer };
