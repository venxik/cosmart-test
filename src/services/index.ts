import { booksApi } from './query';

export const combinedMiddleware = [booksApi.middleware];

export * from './query';
