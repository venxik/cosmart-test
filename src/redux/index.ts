/* eslint-disable @typescript-eslint/no-var-requires */
import { configureStore, PreloadedState } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';

import { rootReducer } from './rootReducer';
import { combinedMiddleware } from '../services';

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(combinedMiddleware),
    preloadedState,
  });
};

const store = setupStore();

type RootState = ReturnType<typeof rootReducer>;
type AppStore = ReturnType<typeof setupStore>;
type AppDispatch = AppStore['dispatch'];

setupListeners(store.dispatch);

const useReduxDispatch = () => useDispatch<AppDispatch>();
const useReduxSelector: TypedUseSelectorHook<RootState> = useSelector;

export { AppStore, RootState, store, useDispatch, useReduxDispatch, useReduxSelector, useSelector };
