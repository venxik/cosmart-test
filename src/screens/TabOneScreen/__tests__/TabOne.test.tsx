import { screen } from '@testing-library/react-native';
import { getBooksResponseMock, renderWithProviders } from '../../../__mocks__';
import * as hooks from '../../../services/query';
import TabOneScreen from '../TabOneScreen';
import { QueryActionCreatorResult } from '@reduxjs/toolkit/dist/query/core/buildInitiate';
import {
  BaseQueryFn,
  FetchArgs,
  FetchBaseQueryError,
  FetchBaseQueryMeta,
  QueryDefinition,
} from '@reduxjs/toolkit/dist/query';
import { SubjectsResponse } from '../../../redux/books';

const spyOnHooks = () => {
  jest.spyOn(hooks, 'useGetLoveSubjectsQuery').mockReturnValue({
    data: getBooksResponseMock,
    isError: false,
    isLoading: false,
    refetch: function (): QueryActionCreatorResult<
      QueryDefinition<
        void,
        BaseQueryFn<string | FetchArgs, unknown, FetchBaseQueryError, {}, FetchBaseQueryMeta>,
        never,
        SubjectsResponse,
        'booksapi'
      >
    > {
      throw new Error('Function not implemented.');
    },
  });
};

describe('TabOneScreen', () => {
  it('match snapshot', async () => {
    spyOnHooks();
    renderWithProviders(<TabOneScreen />);
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
