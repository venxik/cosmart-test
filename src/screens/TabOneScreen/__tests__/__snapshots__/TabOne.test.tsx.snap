// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabOneScreen match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderColor": "#2f95dc",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "authors": Array [
            Object {
              "key": "/authors/OL4327048A",
              "name": "Emily Brontë",
            },
          ],
          "bookingDate": "1234",
          "edition_count": 123,
          "has_fulltext": true,
          "ia": "testing",
          "key": "testing",
          "title": "testing",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
                "borderColor": "#2f95dc",
              },
              Object {
                "borderRadius": 37.5,
                "borderWidth": 1,
                "marginBottom": 37.5,
                "marginHorizontal": 15,
                "padding": 37.5,
              },
            ]
          }
          testID="book_item"
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                  "borderColor": "#2f95dc",
                },
                Object {
                  "marginBottom": 30,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {
                    "fontSize": 45,
                    "fontWeight": "700",
                  },
                ]
              }
            >
              Title
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Array [
                    Object {
                      "fontSize": 37.5,
                      "fontWeight": "300",
                    },
                    Object {
                      "fontFamily": "space-mono",
                    },
                  ],
                ]
              }
            >
              testing
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                  "borderColor": "#2f95dc",
                },
                Object {
                  "marginBottom": 30,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {
                    "fontSize": 45,
                    "fontWeight": "700",
                  },
                ]
              }
            >
              Author
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Array [
                    Object {
                      "fontSize": 37.5,
                      "fontWeight": "300",
                    },
                    Object {
                      "fontFamily": "space-mono",
                    },
                  ],
                ]
              }
            >
              Emily Brontë
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                  "borderColor": "#2f95dc",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {
                    "fontSize": 45,
                    "fontWeight": "700",
                  },
                ]
              }
            >
              Edition Number
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Array [
                    Object {
                      "fontSize": 37.5,
                      "fontWeight": "300",
                    },
                    Object {
                      "fontFamily": "space-mono",
                    },
                  ],
                ]
              }
            >
              123
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                  "borderColor": "#2f95dc",
                },
                Object {
                  "marginTop": 10,
                },
              ]
            }
          >
            <View
              accessibilityRole="button"
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="button_book_item_testing"
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "margin": 8,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Click Here to Book!
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
